src/include/sof/lib/dai.h

```diff
@@ -78,6 +78,7 @@ struct dai_ops {
 	int (*get_hw_params)(struct dai *dai,
 			     struct sof_ipc_stream_params *params, int dir);
 	int (*hw_params)(struct dai *dai, struct sof_ipc_stream_params *params);
+	int (*hw_free)(struct dai *dai);
```

Comment:
I suggest listing the callback by groups and in the order in which they are used. I found out by reverse-engineering the code that set_config is actually called during the prepare stage *after* the hw_params, and it's clearly not evident from the list.

Reply:
> I suggest listing the callback by groups and in the order in which they are used. I found out by reverse-engineering the code that set_config is actually called during the prepare stage _after_ the hw_params, and it's clearly not evident from the list.

I have to admit that I don't know the order the callback are used. Maybe someone can create another PR to do it.

Reply:
well at the minimum you need to make sure the .hw_params is called *after* set_config, otherwise the clocks will be set with values that haven't been initialized yet. Unfortunately in my experiments this was the other way around, set_config was called in the prepare stage *after* hw_params. 

@keyonjie and @lgirdwood we really need your help to clarify what the dai_driver state machine looks like and what callbacks are used in what order.

Reply:
@plbossart set_config is called before hw_params for sure both in the case of static pipelines (during tplg parsing and during BE hw_params) and dynamic pipelines (only during BE hw_params)

Reply:
@ranj063 my traces show otherwise. I think there is a difference between the IPC DAI_CONFIG and the set_config. See the code, the .set_config callback is used in the .prepare stage.

Reply:
@plbossart From my trace log, set_config is called before hw_params. Look into the code, dai_set_config is called when we send a IPC with SOF_IPC_DAI_CONFIG cmd, which is in `sof_link_load`. In short, set_config is called during topology loading stage and hw_params is called during pcm open stage. So set_config should be called before hw_params.